WIP: Add support for the Julia language

IanButterworth opened the pull request:
### What are you trying to accomplish?

Goal: Add support for the Julia language.
Status: Incomplete. Untested. Opening to sanity check progress.

### Anything you want to highlight for special attention from reviewers?

Julia already has https://github.com/JuliaRegistries/CompatHelper.jl but the benefits of consolidating into dependabot are clear.

### How will you know you've accomplished your goal?

I'm new to this codebase and haven't been able to get tests to run locally yet, so I am opening this as WIP to see if CI gets further than I did locally.

### Checklist

<!-- Before requesting review, please ensure that your pull request fulfills the following requirements: -->

- [ ] I have run the complete test suite to ensure all tests and linters pass.
- [ ] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [ ] I have written clear and descriptive commit messages.
- [ ] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [ ] I have ensured that the code is well-documented and easy to understand.

DilumAluthge commented on julia/helpers/Project.toml:
```diff
@@ -0,0 +1,13 @@
+name = "DependabotHelper"
+uuid = "4c4b75a0-81e9-11ee-a506-6b70a314a9a5"
+authors = ["Dependabot <support@dependabot.com>"]
```
```suggestion
authors = ["Dependabot <support@dependabot.com>",
"contributors"]
```

DilumAluthge commented on julia/helpers/src/DependabotHelper.jl:
```diff
@@ -0,0 +1,38 @@
+module DependabotHelper
+
+using JSON
```
```suggestion
import JSON
```

DilumAluthge commented on julia/helpers/src/DependabotHelper.jl:
```diff
@@ -0,0 +1,38 @@
+module DependabotHelper
+
+using JSON
+using Pkg
```
```suggestion
import Pkg
```

DilumAluthge commented on julia/helpers/src/functions.jl:
```diff
@@ -0,0 +1,79 @@
+    temp_dir = mktempdir()
+
+    try
+        # Initialize a project in the temporary directory
+        Pkg.activate(temp_dir)
```
```suggestion
        original_project = Base.active_project()
        # Initialize a project in the temporary directory
        Pkg.activate(temp_dir)
```

DilumAluthge commented on julia/helpers/src/functions.jl:
```diff
@@ -0,0 +1,79 @@
+            "version" => string(pkg_info.version)
+        )
+    finally
+        Pkg.activate()  # Reset to default environment
```
```suggestion
        Pkg.activate(original_project)  # Reset to previous environment
```

DilumAluthge commented on julia/spec/fixtures/julia/registry_responses/E/Example/Compat.toml:
```diff
@@ -0,0 +1,11 @@
+julia = "0.6-1"
+
+["0.5.3-0"]
+julia = "1"
```
Can you add a newline character at the end of this file?

DilumAluthge commented on julia/spec/fixtures/julia/registry_responses/E/Example/Package.toml:
```diff
@@ -0,0 +1,3 @@
+name = "Example"
+uuid = "7876af07-990d-54b4-ab0e-23690620f79a"
+repo = "https://github.com/JuliaLang/Example.jl.git"
```
Can you add a newline character at the end of this file?

IanButterworth commented:
Also @sachin-sandhu is there a way I can test this PR out on an external Julia repo before its merged? i.e. like the way you can use actions from branches, not just released ones

DilumAluthge commented on julia/helpers/Project.toml:
```diff
@@ -0,0 +1,9 @@
+
+[compat]
+DependabotHelper = "0.1.0"
+julia = "1"
```
`[sources]` isn't supported on old Julia versions. Should this compat entry require a newer Julia version?

DilumAluthge commented on julia/helpers/build:
```diff
@@ -0,0 +1,26 @@
+julia --project="$install_dir" -e 'import Pkg; Pkg.instantiate()'
+
+echo "Julia helpers installed successfully in $install_dir"
+echo "Julia depot path: $JULIA_DEPOT_PATH"
```
```suggestion
echo "Julia depot path: ${JULIA_DEPOT_PATH:?}"
```

DilumAluthge commented on julia/CHANGELOG.md:
```diff
@@ -0,0 +1,16 @@
+
+## Unreleased
+
+- Initial support for Julia dependency updates #12316
```
```suggestion
- Initial support for Julia dependency updates (#12316)
```

DilumAluthge commented on julia/helpers/DependabotHelper.jl/src/DependabotHelper.jl:
```diff
@@ -0,0 +1,86 @@
+        else
+            Dict("result" => function_result)
+        end
+    catch err
```
Will the `stderr` output from DependabotHelper.jl show up in the Dependabot logs? If so, it might be helpful to log some info on the exceptions that we catch.

```suggestion
    catch err
        bt = catch_backtrace()
        @info "" exception=(err,bt)
```

DilumAluthge commented on julia/helpers/DependabotHelper.jl/src/functions.jl:
A few initial thoughts on this file:

1. This file is quite large. Would it be possible to split it up into a few smaller files?
2. Will the `stderr` output from DependabotHelper.jl be printed to the Dependabot logs? If so, in this file (the `src/functions.jl` file), I think that it could be helpful, wherever you `catch` a Julia exception, to do something like `@info "" exception=(ex, catch_backtrace())`, to make it a little easier to debug problems.

DilumAluthge commented on julia/helpers/DependabotHelper.jl/src/functions.jl:
```diff
@@ -0,0 +1,1139 @@
+                "manifest_path" => manifest_file_path
+            )
+        finally
+            # Restore original environment
```
This functionality (a `try`-`finally`, in which the original environment is `Pkg.activate`d in the `finally` block) is used in a couple different places. Can you factor it out into a separate function, as such?

```julia
function with_active_project(f::Function)
    original_env = Base.active_project()
    result = try
        f()
    finally
        Pkg.activate(original_env)
    end
    return result
end
```

IanButterworth commented:
@DilumAluthge thanks for the review! Before I work on the suggestions I'll wait to get some feedback from the dependabot-core team to check that the direction here is correct/acceptable.

sachin-sandhu commented:
> @DilumAluthge thanks for the review! Before I work on the suggestions I'll wait to get some feedback from the dependabot-core team to check that the direction here is correct/acceptable.

Hi @IanButterworth , 

good point, we are refactoring the ecosystems right now, Code may benefit from little tweaking. I will provide my review comments by eod today.

sachin-sandhu commented on julia/lib/dependabot/julia/update_checker/latest_version_finder.rb:
```diff
@@ -0,0 +1,87 @@
+        return nil unless latest_version
+
+        # Filter out ignored versions
+        versions = [latest_version]
```
we can inherit from Dependabot::Package::PackageLatestVersionFinder which has standard filters such as filter_ignored_versions , filter_prerelease_versions etc. You can see example here https://github.com/dependabot/dependabot-core/blob/main/elm/lib/dependabot/elm/update_checker/latest_version_finder.rb#L76 . you can override the function in case standard function needs tweaking.

IanButterworth replied:
I think this was made moot by the above change?

robaiken commented:
Hi @IanButterworth! Really excited about this update, it motivated me to start documenting the process for creating new ecosystems and clarifying our guidelines: https://github.com/dependabot/dependabot-core/pull/12477

IanButterworth commented:
My plan is to do a push on this with https://github.com/dependabot/dependabot-core/pull/12477 as reference once its merged.

If there's any other guidance that'd be helpful before that push, please let me know.

IanButterworth commented:
To keep this fresh and not get too far behind main I've updated it and addressed most of the suggestions thus far.

@robaiken I used your guide with Claude to fix issues and create the temporary file summary JULIA_IMPLEMENTATION_STATUS.md

robaiken commented:
> If there's any other guidance that'd be helpful before that push, please let me know.

@IanButterworth we're actually working on improving our docs since they're pretty lacking right now, so any suggestions you have would be awesome.

> I used your guide with Claude to fix issues and create the temporary file summary JULIA_IMPLEMENTATION_STATUS.md

Thanks! Glad the guide was helpful for getting things sorted out with the Julia implementation. If there's anything else you need help with or want me to add to the doc I'm creating, just let me know!

robaiken commented:
@IanButterworth Sometimes during peak times we get rate limited, so we see random failures like this. It's not related to your changes, just timing with the infrastructure. I've merged the new ecosystems documentation, but feel free to reach out with any questions and I'll amend the doc at a later point.

sachin-sandhu reviewed (commented):
@IanButterworth,

Thanks for working on this ecosystem , we have gone through a partial review, you can check the review comments and let us know in case of any concerns. The team will finish the review and let you know shortly.

IanButterworth commented on julia/helpers/DependabotHelper.jl/LICENSE:
My plan is to move this package into its own repo, so this was to prepare for that. I guess I should move the package out sooner.

jeffwidman replied:
I have not looked closely at this PR, is this a core piece of functionality for Dependabot supporting Julia? 

I am leery of having the core functionality live outside of Dependabot repo unless it's owned by the official Julia team. 

It's one thing to have a semi-optional library that we can quickly replace if a maintainer loses interest, but another thing entirely to have the major bits live outside of our control.

There's also security concerns since realistically we don't have the resources to closely audit the source code of our dependencies. Again, if it's under the ownership of the core language/package manager team, that's probably fine.

DilumAluthge replied:
Yes, we can put DependabotHelper.jl in JuliaCI, JuliaLang, or JuliaRegistries. My guess is that JuliaRegistries is the best place, because that's where our registry (https://github.com/JuliaRegistries/General) lives, and where CompatHelper.jl (a Julia-specific Dependabot-like tool that we plan to deprecate in favor of Dependabot) lives.

IanButterworth replied:
Perhaps we could keep it internally here while this is behind the experimental flag, and move it out once it's proven stable?

IanButterworth commented on julia/helpers/DependabotHelper.jl/Project.toml:
```diff
@@ -0,0 +1,26 @@
+version = "0.1.0"
+
+[deps]
+JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
```
Because julia packages must have a UUID as an identifier. https://pkgdocs.julialang.org/v1/toml-files/#The-uuid-field

It allows multiple packages with the same name to exist and not be confused. Name clashes can happen especially if users use multiple registries that don't coordinate name spacing (I use 3 independent registries concurrently)

sachin-sandhu replied:
@IanButterworth , 
makes sense, thanks for clarification.

IanButterworth commented on julia/lib/dependabot/julia/file_updater.rb:
```diff
@@ -0,0 +1,224 @@
+          update_dependency_in_manifest(dependency, parsed_manifest)
+        end
+
+        T.cast(TomlRB.dump(parsed_manifest), String)
```
I've moved away from TomlRB.dump to a regex replace approach. Happy to see how this goes in the wild and adapt as needed.

IanButterworth commented on JULIA_IMPLEMENTATION_STATUS.md:
Reminder to self to remove this file before merge

robaiken commented:
@IanButterworth Do you have any test repos with basic Julia project setups I can use for Dependabot testing? Looking for some starting points with different `Project.toml` configurations

IanButterworth commented:
For a basic start, I'd recommend forking a real package like [CSV.jl](https://github.com/JuliaData/CSV.jl) and modifying the `[compat]` entries to remove newer versions to see if they get noticed by dependabot https://github.com/JuliaData/CSV.jl/blob/04ec1cf3a7b0f069c573ba7c9da351026427532b/Project.toml#L23

honeyankit commented:
@robaiken Approved based on past reviews
